### ISSUE TITLE
pdp11: DEUNA: some questionable code in pdp11_xu.c 

### ISSUE THREAD

**Issue by wfjm** [simh/simh#405]
I'm currently implementing a DEUNA/DELUA interface for my FPGA pdp-11/70, see project GitHub project [w11](https://github.com/wfjm/w11). Even though the back-end code is a complete rewrite I use the `pdp11_xu.c` as reference. When cross checking the DEUNA and DELUA User Guide and the `pdp11_xu.c` sources I stumbled across several sections of code I consider questionable.

**Errors detected in the execution of functions aren't propagated.**
The function handling code in `xu_command` detects errors and returns them as return value, like

      return PCSR0_PCEI;
      return PCSR0_PCEI + 1;            <-- apparently wants to set pr1 PCTO too

but the return value is dropped in only call of `xu_command` 

      case CMD_GETCMD:        /* GET COMMAND */
        xu_command(xu);                      <-- return ignored
        xu->var->pcsr0 |= PCSR0_DNI;         <-- DNI always set !!

I'd expected here something like

      int32 tmp = xu_command(xu);
      xu->var->pcsr0 |= tmp & 0xff00;
      if (tmp & 0x1) xu->var->pcsr1 |= PCSR1_PCTO;

**function RSTAT/RCSTAT doesn't return CURMLT and the DEUNA value for MAXCTR**
The DEUNA UG clearly states that 

         pcbb+4 <15:08>   CURMLT   current number of multicast IDs
         pcbb+4 <07:00>   MAXMLT   maximum number of multicast IDs
         pcbb+6 <15:00>   MAXCTR   maximum length in words for counters 

the code in `pdp11_xu.c` is

      case FC_RSTAT:          /* read extended status */
      case FC_RCSTAT:         /* read and clear extended status */
        value = xu->var->stat;
        wstatus = Map_WriteW(xu->var->pcbb+2, 2, &value);
        value = 10;
        wstatus2 = Map_WriteW(xu->var->pcbb+4, 2, &value);
        value = 32;
        wstatus3 = Map_WriteW(xu->var->pcbb+6, 2, &value);

I'd expect `pcbb+4` to contain `mtlen` in the high byte.
I'd expect `pcbb+6` to be 32 for DEUNA, 34 for DELUA.

**function RDCTR/RDCLCTR returns always 34 words.**
That's imho questionable in multiple ways:
1. number of counters depends on interface type
         DEUNA: last documented value UDBB+76 which is always 0 !! --> 32
         DELUA: last documented value UDBB+102                     --> 34 !!
                UG states that up to 55 can be available (CTRLEN docs)

 2. the first word should according to UG contain the number of words written into UDBB. pdp11_xu.c simply does  `udb[0]  = 68; /* udb length */`.
Shouldn't it return instead the transfered length in words ?
  
3. `pcbb+6` contains CTRLEN, the maximal number of words to transfer
    `pcbb+6` is currently ignored, always 34 words are transfered.

It is also inconsistent to report with RSTAT MAXCTR=32 and send always in RDCTR 34.

**function RMAL seems to return a truncated list**

     case FC_RMAL:   /* read multicast address list */
        mtlen = (xu->var->pcb[2] & 0xFF00) >> 8;
        udbb = xu->var->pcb[1] | ((xu->var->pcb[2] & 03) << 16);
        wstatus = Map_WriteB(udbb, mtlen * 3, (uint8*) &xu->var->setup.macs[2]);
        break;

so transfer size is `mtlen*3` bytes. `mtlen` is defined as '_number of multicast addresses read_'. Shouldn't it be `mtlen*6` bytes transfered ?

**definition of mode bit INTL bit**

     #define MODE_INTL   0000200    /* <07> Internal Loopback [DELUA only] */

but DELUA UG states INTL is bit 6  and bit 7+8 are MBZ.

In normal operation all these sections aren't touched, there are no memory access errors, drivers rarely read back the DEUNA state, and statistic counters maybe rarely used. Only diagnostic code will usually check these things.

It be great of somebody with good testing capabilities could have a look at these points.

Walter

**Comment by larsbrinkhoff** [comment]
I agree in principle that testing every nut and bolt is a good thing.  It's just that resources are limited, and much SIMH work happens in those rare minutes between work, sleep, raising kids, etc.

I don't see any automated testing in this repository.  Adding an automated build with e.g. Travis could be a step forward in the QA department.

**Comment by hrlzm** [comment]
Regarding the first issue, return value from xu_command, in the initial version of the
deuna emulation (see https://www.pdc.kth.se/~bygg/tops/dec_deuna.c) the return
value was the interrupt bits to set, and the virtual bit (1) meant set or clear the PCTO
bit in PCSR1.  Apparently this functionality was done in-line in later versions of the code,
but the return statement still stayed there.

Since I only did the initial code, I have no idea why.

--Johnny

**Comment by dthittner** [comment]
I agree with the main points of the OP that there is questionable code in pdp11_xu.{h,c}.

The first (error return) looks like an artifact of style conversion from a return value to in-line register modification, probably when it was realized that multiple registers would need modification based on the return value, and in-line register modification was selected instead of returning a struct{ }. If the return value is to be truly ignored the function should be declared as a void return, or fixed up to return correct value(s) as noted. IIRC, DNI must always be set, except on an explicit NO-OP.

As for the second point (RSTAT), yes, it appears that the fields are being filled incorrectly. This may be the reason that VAXClusters and LAT are a bit shaky with the DEUNA/DELUA, since they depend on multicast addresses.

Third point (RDCNTR), yes the DEUNA and DELUA have different UDB lengths. I think that I saw the documentation state that UDB[0] is read-only by the controller, so we should respond to the passed length rather than hard-setting it to the DEUNA value. I think the missing DELUA fields at the end would probably be zeros anyway, but diagnostics may do a deeper verification of the structure and complain if they are not there. At the very minimum, it should have an if (DELUA) test to fill in the missing fields.

Fourth item (RMAL), this may or may not be truncating the multicast list. The manual explicitly specifies mtlen*3 (words) [=mtlen*6 (bytes)] as the transfer size, and without checking the parameter of Map_WriteB, this may be an artifact of converting between Map_WriteB and Map_WriteW as is noted in the top comments. It appears to be a bug based on an assumed length of bytes as the passed size parameter to Map_WriteB( ). This multicast issue is also reflected back in point #2.

Fifth item (INTL). Yep, that's a fat finger bug for sure, probably from when the code was bouncing between octal and hex constants. Not sure where that came from :-)

The Q-Bus DEQNA/DELQA/DELQA-T code is much better debugged because so many people use it and report defects. The UNIBUS DEUNA/DELUA code has always had very few bug reports, leading us to assume that it used by far less people. This is the best bug report on the DEUNA/DELUA subsystem in a long time. At some point, it would help to try to run each OS's suite of Digital diagnostics packages for the DEUNA/DELUA.

I'll see if either Mark P or myself have the cycles to do a deeper dive and correct the code.

Dave

**Comment by wfjm** [comment]
@hrlzm @dthittner 
Regarding the first issue (error returns from `xu_command`):
`xu_command` does not do any inline handling of register updates, look at the very beginning

    /* Grab the PCB from the host. */
    rstatus = Map_ReadW(xu->var->pcbb, 8, xu->var->pcb);
    if (rstatus != 0)
      return PCSR0_PCEI + 1;

    /* High 8 bits are defined as MBZ. */
    if (xu->var->pcb[0] & 0177400)
      return PCSR0_PCEI;

In case of error there is right away a return, with either `PCSR0_PCEI` or `PCSR0_PCEI + 1`. So as far as I can see the error condition currently is simply lost.
This got undetected because it's irrelevant in practice, there will simply be never a memory access error in a normal operating system environment.

**Comment by wfjm** [comment]
@dthittner wrote

> IIRC, DNI must always be set, except on an explicit NO-OP.

The DEUNA UG is unambiguously clear that **either** `DNI` in case of success **or** `PCEI` in case of failure is set, see

    DEUNA UG p 4-5
      The DEUNA responds by executing the port command and setting the Done
      Interrupt (DNI) *or* the Port Command Error Interrupt (PCIE) bits.

      The DEUNA notifies the host of completion of the Port Command via
      interrupt; *either* Done Interrupt (DNI) for successful completion
      *or* Port Command Error Interrupt (PCIE) for failure.

    DEUNA UG 4-16
      Command execution ends with the setting of *either* the DNI *or* the
      PCEI bits in PCSR0.

    DEUNA UG 4-65  Figure 4-29 shows the DNI/PCEI either/or graphically.
